Added a guide to deploy on bare-metal/VMs (#51085)
keycloak/keycloak: docs/guides/high-availability/multi-cluster-v2/deploy-keycloak-bare-metal.adoc

* TLS certificate and private key files for HTTPS.
* A load balancer in front of the {project_name} nodes.
* The database CA certificate file for securing the database connection.
* Network connectivity between all {project_name} nodes within each site for cache clustering and failure detection (TCP ports 7800 and 57800 by default).
See the <@links.server id="caching" anchor="network-ports"/> {section} for details.

== Load balancer topology

For a multi-site deployment, two load balancer topologies are common:

Per-site load balancer with a global load balancer in front::
Each site has its own load balancer (for example, HAProxy or NGINX) that distributes traffic across the {project_name} nodes within that site, probing `/health/ready` on management port 9000 of each node.
A global load balancer routes client traffic to the per-site load balancers, probing `/lb-check` through them to detect site-level failures.
+
This topology provides better operational isolation: individual sites can be drained or taken offline for maintenance without affecting the global load balancer configuration.

> It also improves security, as TLS can be terminated at the site level and site-specific policies can be enforced, limiting the blast radius of a compromise.

Reviewer: I'd say this is debatable, and different people can have different opinions here. 

For example: 

* TLS passthrough might be frowned upon as inspection by a web application firewall is not possible. Maybe you want to block harmful traffic as early as possible, and not at the site level 
* Site-specific policies are possible, but an asymmetric configurations can have risks of misconfigurations. Usually I would recommend them being symmetric for the sake of simplicity. 

With that, I'd recommend to remove that sentence.

Single load balancer::
One load balancer routes directly to all {project_name} nodes across all sites, probing `/health/ready` on management port 9000 of each node.
+

> This topology is simpler to manage, but its shared load-balancing tier has a larger failure domain and must be deployed redundantly to avoid becoming a single point of failure.

Reviewer: The loadbalancers at the front should be highly available in both setups, so this is not a differentiator. With that, I'd suggest to remove this sentence.

Per-site operational tasks such as draining a site for maintenance or isolating a failing site require reconfiguring a shared component, which increases operational risk.

> For production deployments, the per-site topology with a global load balancer is recommended.

Reviewer: If those two sentences are removed, there might be fewer reasons to do this. So this sentence might become a bit weaker like:

```suggestion
The following example assumes a per-site topology with a global load balancer.
```

> With the `stateless` feature, {project_name} does not store session data in memory, so configuring sticky sessions on the load balancer is not required.

Reviewer: It was never required to configure sticky sessions in any of the LB setups, it was just a recommendation to better leverage the caches. 

My suggestion:

```suggestion
With the `stateless` feature, {project_name} does not store session data in memory, so configuring sticky sessions on the load balancer is no longer recommended as it will not provide a performance benefit.
```

We are still caching realm data, so a warmed-up realm cache could provide some benefits. But to me this doesn't outweigh the downsides of an imbalanced load, so I'd stick to the slight simplification I suggested above.

Adjust this number to meet the needs of your system.
With the `stateless` feature, authentication sessions and action tokens are stored in the database, which increases database load compared to setups with an external {jdgserver_name} cluster. Adjust the pool size accordingly.
See the <@links.ha id="multi-cluster-concepts-database-connections" /> {section} for details.
<3> Enable the `stateless` feature to remove the requirement for an external {jdgserver_name} cluster.
<4> Set a unique cluster name that identifies this cluster. Each cluster must have a distinct name; no two clusters may share the same name.
This option is only available if the `stateless` feature is enabled.
<5> Set the hostname clients use to connect to {project_name}. This should match the hostname configured on the load balancer.
<6> Provide the TLS certificate and private key files for HTTPS.
<7> Secure the database connection. Specify the path to the database CA certificate file. {project_name} automatically loads certificates from the specified truststore path.
<8> Enable the PROXY protocol so that {project_name} can read the real client IP address from the PROXY protocol header sent by the load balancer.

> The load balancer must also be configured to send PROXY protocol headers (for example, `send-proxy-v2` in HAProxy).

Reviewer: ```suggestion
To use this option, the load balancer must also be configured to send PROXY protocol headers (for example, `send-proxy-v2` in HAProxy). 
If your load balancer does not support the PROXY protocol, remove this option to use plain HTTPS.
```

If we think that configuring the PROXY protocol makes things too complicated/too specific, we can skip it here, and just add a pointer to the docs how to add it.

Do not set `--proxy-headers` when using TLS passthrough; the two options are mutually exclusive.
See the <@links.server id="haproxy-passthrough"/> or <@links.server id="traefik-passthrough"/> {sections} for complete examples.
The `--shutdown-delay` keeps {project_name} running long enough for the load balancer to detect the shutdown via health check failures and remove the node from rotation.
Set this value to at least the load balancer's detection time (health check interval multiplied by the failure threshold).
See the Graceful HTTP shutdown section in the <@links.server id="reverseproxy"/> {section} for details.
<9> To be able to analyze the system under load, enable the metrics endpoint.
<10> Limit the number of queued requests for load shedding. All exceeding requests are served with an HTTP 503.
See the <@links.ha id="multi-cluster-concepts-threads" anchor="multi-cluster-load-shedding"/> {section} about load shedding for details.

[#multi-cluster-v2-bare-metal-verify-deployment]
== Verify the deployment

Confirm that each {project_name} node is ready by checking the health endpoint on the management port:

[source,bash]
----

> curl -sk https://localhost:9000/health/ready

Reviewer: Nitpick: This command will not show the HTTP status code that you mention in the paragraph below. For that, you'd need to AFAIK add `-v` as well.

----

A healthy node returns HTTP 200 with an overall status of `UP`.
The `checks` array contains the readiness checks enabled for the deployment.

> Repeat this check for each node in each site.

Reviewer: For completeness: Add a note to check the `/lb-check` per site returns `200 OK/UP`.

<#import "/templates/guide.adoc" as tmpl>
<#import "/templates/links.adoc" as links>

<@tmpl.guide
title="Deploying {project_name} for HA on bare-metal or virtual machines (v2)"
summary="Deploy {project_name} for multi-cluster HA (stateless) on bare-metal or virtual machines as a building block."
tileVisible="false"
preview="true"
previewDiscussionLink="https://github.com/keycloak/keycloak/discussions/50640" >

This {section} describes advanced {project_name} configurations for bare-metal servers or virtual machines that will recover from single node failures or site failures.

> The concepts and configuration options are equivalent to the {kubernetes}-based blueprint, which is load tested as described in the <@links.ha id="multi-cluster-v2-introduction" anchor="multi-cluster-v2-tested-configuration"/> {section}.

Reviewer 2: I don't think we want Kubernetes to be parameterised here as the guide name never changes upstream or downstream, instead we should link to the guide using `<@links.ha ...>`.

Reviewer 2: Ignore if you remove this line as suggested on my comment below.

You should verify the performance and failure behavior of your specific environment.

These instructions are intended for use with the setup described in the <@links.ha id="multi-cluster-v2-concepts"/> {section}.

> This is an alternative to the <@links.ha id="multi-cluster-v2-deploy-keycloak-kubernetes" /> {section} for environments that do not use {kubernetes}.

Reviewer 2: It seems we're repeating what we already say on L12 here. I would condense this section to three points:

1.  Purpose of doc (L11)
2. L15
3. Performance disclaimer (L13)

== Prerequisites

* https://openjdk.java.net/[OpenJDK {jdk_version_latest}] installed on each node.
* {archivedownloadurl}[{archivebasename}-{version}.zip] downloaded and extracted on each node.
* A highly available database deployed with synchronous replication across sites, reachable from all {project_name} nodes. The following example assumes a PostgreSQL database. Any database supported by {project_name} will work.
* TLS certificate and private key files for HTTPS.
* A load balancer probing `/health/ready` on management port 9000 of each {project_name} node.

> In the per-site topology, the global load balancer should probe `/lb-check` through each per-site load balancer.

Reviewer 2: I think in the Prerequisites section we should just state that a load balancer is required and then describe how it should be configured in the Load Balancer section that comes later.

* The database CA certificate file for securing the database connection.

> * Network connectivity between all {project_name} nodes within each site on TCP ports 7800 and 57800 (or their configured equivalents) for cache clustering and failure detection.

Reviewer 2: I would remove "or their configured equivalents" to be more concise, how about:

"Network connectivity between all {project_name} nodes within each site for cache clustering and failure detection (TCP ports 7800 and 57800 by default)."

See the <@links.server id="caching" anchor="network-ports"/> {section} for details.

> The following setup assumes a TLS passthrough setup, though this is not limited to this configuration.

Reviewer 2: I'm not sure this belongs here.

Each site has its own load balancer (for example, HAProxy or NGINX) that distributes traffic across the {project_name} nodes within that site.
A global load balancer routes client traffic to the per-site load balancers, probing `/lb-check` through them to detect site-level failures.
+
This topology provides better operational isolation: individual sites can be drained or taken offline for maintenance without affecting the global load balancer configuration.
It also improves security, as TLS can be terminated at the site level and site-specific policies can be enforced, limiting the blast radius of a compromise.

Single load balancer::
One load balancer routes directly to all {project_name} nodes across all sites.
+
This topology is simpler to manage, but its shared load-balancing tier has a larger failure domain and must be deployed redundantly to avoid becoming a single point of failure.
Per-site operational tasks such as draining a site for maintenance or isolating a failing site require reconfiguring a shared component, which increases operational risk.

For production deployments, the per-site topology with a global load balancer is recommended.

> With the `stateless` feature, {project_name} does not store session data in memory, so sticky sessions (session affinity) on the load balancer are not required.

Reviewer 2: ", so configuring sticky session on the load balancer is not required."

. Set the database credentials using environment variables to avoid exposing them in shell history or the process list:
+
[source,bash]
----
export KC_DB_USERNAME=<DB_USERNAME>
export KC_DB_PASSWORD=<DB_PASSWORD>
----

. Start {project_name} with the following command, replacing the placeholder values with your environment-specific settings:
+
[source,bash]
----
bin/kc.sh start \
    --db=postgres \
    --db-url=jdbc:postgresql://<DATABASE_URL>:5432/keycloak \
    --db-pool-initial-size=30 \
    --db-pool-min-size=30 \
    --db-pool-max-size=30 \
    --features=stateless \
    --spi-cache-embedded--default--cluster-name=<CLUSTER_NAME> \
    --hostname=<KEYCLOAK_HOSTNAME> \
    --https-certificate-file=/path/to/tls.crt \
    --https-certificate-key-file=/path/to/tls.key \
    --db-tls-mode=verify-server \
    --truststore-paths=/path/to/db-ca.pem \
    --proxy-protocol-enabled=true \
    --shutdown-delay=30 \
    --health-enabled=true \
    --metrics-enabled=true \
    --event-metrics-user-enabled=true \
    --log-console-output=json \
    --http-max-queued-requests=1000
----
+

> The following list explains the key options:

Reviewer 2: Can we use callouts here like in the other guides?

Set this value to at least the load balancer's detection time (health check interval multiplied by the failure threshold).
See the Graceful HTTP shutdown section in the <@links.server id="reverseproxy"/> {section} for details on tuning this value.
`--health-enabled`::
Enable the health check endpoints so that load balancers and monitoring can verify that {project_name} is ready.
`--metrics-enabled`::
To be able to analyze the system under load, enable the metrics endpoint.
`--http-max-queued-requests`::
Limit the number of queued requests for load shedding. All exceeding requests are served with an HTTP 503.
See the <@links.ha id="multi-cluster-concepts-threads" anchor="multi-cluster-load-shedding"/> {section} about load shedding for details.

[#multi-cluster-v2-bare-metal-verify-deployment]
== Verify the deployment

Confirm that each {project_name} node is ready by checking the health endpoint on the management port:

[source,bash]
----
curl -sk https://localhost:9000/health/ready
----

A healthy node returns HTTP 200 with an overall status of `UP`.
The `checks` array contains the readiness checks enabled for the deployment.

Repeat this check for each node in each site.

> Afterward, verify that the external load balancer routes traffic to both sites by accessing the {project_name} hostname configured on the load balancer.

Reviewer 2: Instead of "afterward", maybe "finally" or "lastly"?

The steps described here would only verify that the hostname resolves to a keycloak instance, not all instances. I think the only way to verify that all nodes are routed to would be to ensure that only one instance is up at a time and cycle through them. Maybe I'm missing something though 🤔
